Updates Fungible Token tutorial to use a different structure and explainations (#1238)
onflow/cadence: docs/tutorial/02-hello-world.mdx

- The `storage` domain is where all objects (such as structs and resource objects representing tokens or NFTs) are stored.
  It is only directly accessible by the owner of the account.
- The `private` domain is like a private API. You can optionally create [links](/cadence/language/capability-based-access-control/)
  to any of your stored assets here.
  Only the owner and anyone they give access to can use these interfaces to call functions that are defined in their stored assets.

> - The `public` domain is kind of like your account's public API. It is exactly like the `private` domain in that

Reviewer: ```suggestion
- The `public` domain is your account's public API. It is exactly like the `private` domain in that
```

> - The `public` domain is your account's public API. It is exactly like the `private` domain in that
>   the owner can link capabilities here that point to stored assets.
>   Anyone in the network can access the `public` domain's functionality.

Reviewer 2: ```suggestion
- The `public` domain is your account's public API. It is like the `private` domain in that
  only the owner can link capabilities here that point to stored assets.
  The difference is that anyone in the network can access the `public` domain's functionality.
```

>   The difference is that anyone in the network can access the `public` domain's functionality.

Reviewer: ```suggestion
  Anyone in the network can access the `public` domain's functionality.
```

Transactions are divided into two main phases, `prepare` and `execute`.

> 1. The `prepare` phase is required, but is not used here, so it will be covered later.

Reviewer: ```suggestion
1. The `prepare` phase is required, but is not used here. We'll cover this phase later on.
```

This contract that we just deployed declares a definition for the `HelloAsset` resource.

> Let's walk through this contract, starting with the resource, one of the most important things

Reviewer: ```suggestion
Let's walk through this contract, starting with the resource. Resources are one of the most important things
```

```

Resources are a composite type similar to a struct or a class because they can have any number of fields or functions within them.
The difference is how code is allowed to interact with them.

- Each instance of a resource can only exist in exactly one location and cannot be copied.
  Here, location refers to account storage, a temporary variable in a function, a storage field in a contract, etc.
- Resources must be explicitly moved from one location to another when accessed.
- Resources also cannot go out of scope at the end of a function execution.
  They must be explicitly stored somewhere or destroyed.

These characteristics make it impossible to accidentally lose a resource from a coding mistake.

Resources are useful when you want to model direct ownership.
Traditional structs or classes from other conventional programming languages

> are not an ideal way to represent this ownership because they can be copied.

Reviewer: ```suggestion
are not an ideal way to represent ownership because they can be *copied*.
```

This transaction imports the `HelloWorld` definitions from the account we just deployed them to, removes the `HelloAsset` object from storage,
and calls the `hello()` function of the stored `HelloAsset` resource.

This transaction also is our first to use the `prepare` phase! 
The `prepare` phase is the only place that has access to the signing accounts' [private `AuthAccount` object](/cadence/language/accounts/#authaccount).
`AuthAccount` has special methods that allow saving to and loading from `/storage/`,

> and creating `/private/` and `/public/` links to the objects in `/storage/`.
> (called [capabilities](/cadence/language/capability-based-access-control), will be covered later)

Reviewer: ```suggestion
and creating `/private/` and `/public/` links to the objects in `/storage/`, called [capabilities](/cadence/language/capability-based-access-control) (more on these later).
```

The `execute` phase does not have access to `AuthAccount` and thus can only use objects
that were removed from the `AuthAccount` in the `prepare` phase or created in the `prepare` phase.

By not allowing the execute phase to access account storage,
we can statically verify which assets and areas of the signers storage a given transaction can modify.
Browser wallets and applications that submit transactions for users can use this to show what a transaction could alter,

> and users can have more confidence that they aren't getting fed

Reviewer: ```suggestion
giving users information about transactions that wallets will be executing for them, and confidence that they aren't getting fed
```

Reviewer: Meh, something didn't feel right here, but take-or-leave this one imo.
